tests/unit/test_tiler.py

```diff
@@ -29,18 +30,41 @@
 
 
 class Describe_RandomTiler:
-    @pytest.mark.parametrize("level", (2, -2))
-    def it_constructs_from_args(self, level, request):
+    @pytest.mark.parametrize("level, mpp", ((2, None), (-2, None), (2, 0.5)))
+    def it_constructs_from_args(self, level, mpp, request):
         _init = initializer_mock(request, RandomTiler)
 
-        random_tiler = RandomTiler((512, 512), 10, level, 7, True, "", ".png", int(1e4))
+        random_tiler = RandomTiler(
+            (512, 512), 10, level, 7, True, "", ".png", int(1e4), mpp=mpp
+        )
 
         _init.assert_called_once_with(
-            ANY, (512, 512), 10, level, 7, True, "", ".png", int(1e4)
+            ANY, (512, 512), 10, level, 7, True, "", ".png", int(1e4), mpp=mpp
         )
         assert isinstance(random_tiler, RandomTiler)
         assert isinstance(random_tiler, Tiler)
 
+    @pytest.mark.parametrize(
+        "level, mpp, expected_level, expected_mpp",
+        ((2, None, 2, None), (None, 0.5, 0, 0.5), (2, 0.5, 0, 0.5)),
+    )
+    def mpp_supercedes_level(self, level, mpp, expected_level, expected_mpp):
+
+        tiler = RandomTiler((512, 512), 10, level=level, mpp=mpp)
+
+        assert tiler.level == expected_level
+        assert tiler.mpp == expected_mpp
+
+    @pytest.mark.parametrize(
+        "mpp, fixed_tile_size", ((None, (512, 512)), (0.5, (1024, 1024)))
+    )
+    def it_can_fix_tile_size_if_mpp(self, mpp, fixed_tile_size):
+        fake_slide = namedtuple("fake_slide", ["base_mpp"])
+        tiler = RandomTiler((512, 512), 0, True, 80, 0, "", ".png", mpp=mpp)
+        tiler._set_proper_tile_size(fake_slide(0.25))
```

Comment:
It took me 5 minutes to understand what was going on here.. 😅 You're faking the value of a property of `Slide` creating the `namedtuple`. IMHO it would be better to use a Slide mock 

However, and more importantly, this is a unit test that is calling `_set_proper_tile_size` and not mocking the internal method that gets called (`_get_proper_tile_size`). Therefore, it's not really testing in isolation. My suggestion would be to have:
1. a unit test that calls and tests the return value of `_get_proper_tile_size`
2. and separately a test for `_set_proper_tile_size` that is basically mocking `_get_proper_tile_size` and testing that the mock is called with the right parameters

Let me know if you have questions or comments 🙏